tModBuildTasks/SynchronizeDirectories.cs

```diff
@@ -0,0 +1,94 @@
+﻿using System;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Build.Framework;
+
+namespace tModLoader.BuildTasks;
+
+/// <summary>
+/// Copies contents from one folder to another, additionally removing entries in the destination that aren't present in the source.
+/// </summary>
+public class SynchronizeDirectories : TaskBase
+{
+	[Required]
+	public string Source { get; set; } = string.Empty;
+
+	[Required]
+	public string Destination { get; set; } = string.Empty;
+
+	protected override void Run()
+	{
+		Source = Path.GetFullPath(Source);
+		Destination = Path.GetFullPath(Destination);
+
+		var source = new DirectoryInfo(Source);
+		var destination = new DirectoryInfo(Destination);
+
+		if (!source.Exists) {
+			Log.LogError($"Source directory '{Source}' doesn't exist!");
+			return;
+		}
+
+		destination.Create();
+
+		// Delete files that shouldn't be there.
+		Parallel.ForEach(destination.EnumerateFiles("*", SearchOption.AllDirectories), file => {
+			string relativePath = IOUtils.SubstringToRelativePath(file.FullName, Destination);
+			string sourcePath = Path.Combine(Source, relativePath);
+
+			if (!File.Exists(sourcePath)) {
+				IOActionWrapper(file.Delete);
+			}
+		});
+
+		// Delete directories that shouldn't be there.
+		Parallel.ForEach(destination.EnumerateDirectories("*", SearchOption.AllDirectories), directory => {
+			string relativePath = IOUtils.SubstringToRelativePath(directory.FullName, Destination);
+			string sourcePath = Path.Combine(Source, relativePath);
+
+			if (!Directory.Exists(sourcePath)) {
+				try {
+					IOActionWrapper(() => directory.Delete(recursive: true));
+				}
+				catch (DirectoryNotFoundException) { }
+			}
+		});
+
+		// Copy files that should be there.
+		Parallel.ForEach(source.EnumerateFiles("*", SearchOption.AllDirectories), sourceFile => {
+			string relativePath = IOUtils.SubstringToRelativePath(sourceFile.FullName, Source);
+			var destinationFile = new FileInfo(Path.Combine(Destination, relativePath));
+
+			if (!IOUtils.AreFilesSeeminglyTheSame(sourceFile, destinationFile)) {
+				IOActionWrapper(() => {
+					Directory.CreateDirectory(destinationFile.DirectoryName);
+					sourceFile.CopyTo(destinationFile.FullName, overwrite: true);
+				});
+			}
+		});
+	}
+
+	// Runs an action with a fixed amount of retries for the case of IOExceptions occuring from third-party short-lived file locks.
+	private void IOActionWrapper(Action action)
```

Comment:
Recommend:
- `RunIOActionWithRetries`
- `DelayMs` of `100`, or a sequence of `100, 200, 400, 800, 1600`